File: kafka-producer/src/main/java/com/integration/agromonitor/kafkaproducer/service/drone/Drone.java
package com.integration.agromonitor.kafkaproducer.service.drone;

import com.integration.agromonitor.kafkaproducer.service.sensor.Sensor;

import java.util.Random;

public class Drone {

    public Sensor sensor;
    final Double LATITUDE_CENTRO_BR = -10.18311;
    final Double LONGITUDE_CENTRO_BR = -48.33366;
    final Double RAIO_OPERACAO = 1000.0 * 1200.0;
    private Boolean tracker = true;

    public String name;
    public Double latitude;
    public Double longitude;

    public Double getLatitude() {
        return latitude;
    }
    public Double getLongitude() {
        return longitude;
    }
    public String getName() {
        return name;
    }

    public String getTracker() {
        return tracker? "ON" : "OFF";
    }

    public Drone(String nameDrone) {
        sensor = new Sensor();
        name = nameDrone;
    }

    public void setTracker() {
        tracker = !tracker;
    }

    public void calculateRandomLocation() {
        Double y0 = LATITUDE_CENTRO_BR;
        Double x0 = LONGITUDE_CENTRO_BR;

        // de metros para graus
        Double radiusInDegrees = RAIO_OPERACAO / 111320f;

        // distancia randomica a partir do eixo.
        Random random = new Random();
        Double u = random.nextDouble();
        Double v = random.nextDouble();
        Double w = radiusInDegrees * Math.sqrt(u);
        Double t = 2 * Math.PI * v;

        // deltas de x e y..
        Double x = w * Math.cos(t);
        Double y = w * Math.sin(t);

        // Compensação de X!
        Double new_x = x / Math.cos(Math.toRadians(y0));

        latitude = (y0 + y);
        longitude = (x0 + new_x);
    }

}

File: src/main/kotlin/DroneControl/Drone.kt
package DroneControl

import SensorControl.MonitorSensor
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.random.Random

/*
  TODO: CRIAR NOME SEQUENCIAL DO DRONE;
 */

class Drone {
    val monitor = MonitorSensor()
    var tracker: Boolean = true

    fun getStatus(): String = if (tracker) "ON" else "OFF"

    fun readValues(): String {
        var s = "■ DRONE #{${this.hashCode()}} >> Tracker: ${getStatus()}"
        if (tracker) {
            var LATITUDE_CENTRO_BR = -10.18311
            var LONGITUDE_CENTRO_BR = -48.33366
            var RAIO_OPERACAO = 1000.0 * 1200.0
            val (latitude, longitude) = getRandomLocation(LATITUDE_CENTRO_BR,  LONGITUDE_CENTRO_BR, RAIO_OPERACAO)
            s += (" * Latitude: {$latitude} - Longitude: {$longitude} => {${monitor.readValues()}}")
        }
        return s
    }

    fun getRandomLocation(latitude: Double, longitude: Double,  raio: Double): Pair<Double, Double> {
        val y0: Double = latitude
        val x0: Double = longitude

        // de metros para graus
        val radiusInDegrees: Double = raio / 111320f

        // distancia randomica a partir do eixo.
        val u = Random.nextDouble()
        val v = Random.nextDouble()
        val w = radiusInDegrees * sqrt(u)
        val t = 2 * Math.PI * v

        // deltas de x e y..
        val x = w * cos(t)
        val y = w * sin(t)

        // Compensação de X!
        val new_x = x / cos(Math.toRadians(y0))

        return (y0 + y) to (x0 + new_x)
    }

}
